### PR TITLE
Kubevirt: Add guest-to-request memory headroom to virt-launchers

### PR DESCRIPTION
**Pull request by iholder101**
**Overview**
This PR introduces a mechanism to add guest-to-request memory headroom (a.k.a. "wiggle-room") to virt-launcher pods.

Please note that the additional memory is being added solely to the virt-launcher pod while the the VMI object remains untouched. This means that the guest memory will remain the same as defined at the VMI level (either explicitly through `vmi.spec.domain.memory.guest` or implicitly by `vmi.spec.domain.resources.requests[memory]`). If the VMI sets limits, the mechanism would skip adding memory headroom.

**Motivation**
Currently, Kubevirt's [overhead calculation](https://github.com/kubevirt/kubevirt/blob/v0.59.0-rc.1/pkg/virt-controller/services/renderresources.go#L272) [1] is not accurate. Therefore, during memory bursts that cause a high memory pressure, VMIs are being OOM killed.

This mechanism ensures that less VMIs are being scheduled to a node, therefore would have a lot of free space on the node that could be used by VMIs during bursts. We know from empiric testings that this dramatically reduces the probability of workloads being killed.

In the future, we plan to improve the memory overhead calculation in Kubevirt. Once the memory overhead calculation is accurate enough and we'd get to a more stable situation, this mechanism will be deleted.

[1] https://github.com/kubevirt/kubevirt/blob/v0.59.0-rc.1/pkg/virt-controller/services/renderresources.go#L272

**How to use this mechanism**
The following two HCO annotations would control the mechanism:
* `kubevirt.io/enable-guest-to-request-memory-headroom: "true"` needs to be used in order to enable this mechanism.
* `kubevirt.io/custom-guest-to-request-memory-headroom: <quantity>` overrides the default memory headroom amount, which is `2GB`.

As an example, let's add the enablement annotation to the HCO object:
```yaml
apiVersion: hco.kubevirt.io/v1beta1
kind: HyperConverged
metadata:
  annotations:
    kubevirt.io/enable-guest-to-request-memory-headroom: "true"
```

Then, we can create vmi-fedora (this is not the full manifest):
```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachineInstance
metadata:
  name: vmi-fedora
spec:
  domain:
    resources:
      requests:
        memory: 1024M
```

The virt-launcher pod then will be created as the following (this is, again, not the full manifest):
```yaml
apiVersion: v1
kind: Pod
spec:
  containers:
-   name: compute
    resources:
      limits:
        devices.kubevirt.io/kvm: "1"
        devices.kubevirt.io/tun: "1"
        devices.kubevirt.io/vhost-net: "1"
      requests:
        cpu: 100m
        devices.kubevirt.io/kvm: "1"
        devices.kubevirt.io/tun: "1"
        devices.kubevirt.io/vhost-net: "1"
        ephemeral-storage: 50M
        memory: "3279755392"  # <-- 1024M + 2GB
```

Then, if we check how the memory size from the guest's perspective:
```bash
$> virtctl console vmi-fedora
[fedora@vmi-fedora ~]$ free -h

               total        used        free      shared  buff/cache   available
Mem:           912Mi       154Mi       601Mi       0.0Ki       156Mi       622Mi
Swap:          911Mi          0B       911Mi
```

**Enforce limits & requests by delta**
As a follow-up to this PR: https://github.com/kubevirt/hyperconverged-cluster-operator/pull/2206, I add two more knobs to allow configuring a memory/cpu delta, not only ratios. The tests and documentation is also updated.

**Reviewer Checklist**

> Reviewers are supposed to review the PR for every aspect below one by one. To check an item means the PR is either "OK" or "Not Applicable" in terms of that item. All items are supposed to be checked before merging a PR. 

- [X] PR Message
- [X] Commit Messages
- [X] How to test
- [X] Unit Tests
- [ ] Functional Tests
- [X] User Documentation
- [X] Developer Documentation
- [X] Upgrade Scenario
- [X] Uninstallation Scenario
- [X] Backward Compatibility
- [X] Troubleshooting Friendly
  
**Release note**:

```release-note
Kubevirt: Add guest-to-request memory headroom to virt-launchers
```

### CODE REVIEW

**Comment by iholder101** [comment]
make sanity passes locally

**Comment by fabiand** [comment]
Without much review, but @iholder101 the term "system-reserved" is used by Kube in a well defined way. Because we are speaking about different memory in this PR, I'd suggest to replace the term (in this PR) "system-reserved" with a different one, i.e. `additional-headroom` or `memory-to-wiggle` (in the latter case please take care to avoid the typo `-to-giggle`).

**Comment by fabiand** [comment]
/hold

The term "system-reserved" needs to be changed to avoid confusion.

**Comment by iholder101** [comment]
> Without much review, but @iholder101 the term "system-reserved" is used by Kube in a well defined way. Because we are speaking about different memory in this PR, I'd suggest to replace the term (in this PR) "system-reserved" with a different one, i.e. `additional-headroom` or `memory-to-wiggle` (in the latter case please take care to avoid the typo `-to-giggle`).

Done!
The new annotation keys are `*-guest-to-request-memory-headroom`.
I've also added the option to enforce limits by deltas, not only via ratios.

PR is now ready to review.
/unhold

**Comment by tiraboschi** [comment]
Can you please also add a functional test under tests/func-tests ?

**Comment by iholder101** [comment]
> Can you please also add a functional test under tests/func-tests ?

Hey,

I was running `make cluster-up && make cluster-sync && make build-functests && make functests`, but faced the following error:
```
++ kubectl -n kubevirt-hyperconverged get pod -l name=hyperconverged-cluster-operator -o 'jsonpath={.items[0] .spec .containers[?(@.name=="hyperconverged-cluster-operator")] .image}'
error: error executing jsonpath "{.items[0] .spec .containers[?(@.name==\"hyperconverged-cluster-operator\")] .image}": Error executing template: array index out of bounds: index 0, length 0. Printing more information for debugging the template:
    template was:
        {.items[0] .spec .containers[?(@.name=="hyperconverged-cluster-operator")] .image}
    object given to jsonpath engine was:
        map[string]interface {}{"apiVersion":"v1", "items":[]interface {}{}, "kind":"List", "metadata":map[string]interface {}{"resourceVersion":""}}


+ operator_image=
make: *** [Makefile:152: test-functional-in-container] Error 1
```

When I run this command manually, everything works fine:
```
iholder_zeus[HCO]> k -n kubevirt-hyperconverged get pod -l name=hyperconverged-cluster-operator -o 'jsonpath={.items[0] .spec .containers[?(@.name=="hyperconverged-cluster-operator")] .image}'
selecting docker as container runtime
registry:5000/kubevirt/hyperconverged-cluster-operator:latest
```

Is it possible to run functional tests locally? How are other doing it?
Am I missing something here?

Thanks

**Comment by iholder101** [comment]
> > Can you please also add a functional test under tests/func-tests ?
> 
> Hey,
> 
> I was running `make cluster-up && make cluster-sync && make build-functests && make functests`, but faced the following error:
> 
> ```
> ++ kubectl -n kubevirt-hyperconverged get pod -l name=hyperconverged-cluster-operator -o 'jsonpath={.items[0] .spec .containers[?(@.name=="hyperconverged-cluster-operator")] .image}'
> error: error executing jsonpath "{.items[0] .spec .containers[?(@.name==\"hyperconverged-cluster-operator\")] .image}": Error executing template: array index out of bounds: index 0, length 0. Printing more information for debugging the template:
>     template was:
>         {.items[0] .spec .containers[?(@.name=="hyperconverged-cluster-operator")] .image}
>     object given to jsonpath engine was:
>         map[string]interface {}{"apiVersion":"v1", "items":[]interface {}{}, "kind":"List", "metadata":map[string]interface {}{"resourceVersion":""}}
> 
> 
> + operator_image=
> make: *** [Makefile:152: test-functional-in-container] Error 1
> ```
> 
> When I run this command manually, everything works fine:
> 
> ```
> iholder[HCO]> k -n kubevirt-hyperconverged get pod -l name=hyperconverged-cluster-operator -o 'jsonpath={.items[0] .spec .containers[?(@.name=="hyperconverged-cluster-operator")] .image}'
> selecting docker as container runtime
> registry:5000/kubevirt/hyperconverged-cluster-operator:latest
> ```
> 
> Is it possible to run functional tests locally? How are other doing it? Am I missing something here?
> 
> Thanks

Hey @tiraboschi,

In addition to what I wrote above, seems that the tests are being skipped if we're not running on Openshift.
Is that a real constraint? How is it possible to run tests locally if so?

**Comment by nunnatsa** [comment]
This PR creates a new coupling between HCO and VMIs. This is not HCO job to deal with VMIs and so it's not the best system design. This logic is more related to KubeVirt itself.

Each coupling make it harder to maintain the software and may create potential risks.

If this is a temporary solution, I would like to see a plan for the long range solution in KubeVirt before approving this PR.

/hold

**Comment by nunnatsa** [comment]
> This is supposed to be a temporary solution that would be deleted in the future.
> The Kubevirt implementation is already ongoing: https://github.com/kubevirt/kubevirt/pull/9322.

:+1: 

Please add these two lines to the PR description.

File: tests/func-tests/virt-launcher-mutating-webhook.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+	cli, err := kubecli.GetKubevirtClientFromRESTConfig(virtCli.Config())
+	Expect(err).ToNot(HaveOccurred())
+
+	res := cli.RestClient().Put().
+		Resource("hyperconvergeds").
+		Name(hcov1beta1.HyperConvergedName).
+		Namespace(flags.KubeVirtInstallNamespace).
+		AbsPath("/apis", hcov1beta1.SchemeGroupVersion.Group, hcov1beta1.SchemeGroupVersion.Version).
+		Timeout(10 * time.Second).
+		Body(hc).Do(context.TODO())
```
Comment (nunnatsa) [inline tests/func-tests/virt-launcher-mutating-webhook.go:120]: This is failing for not having the kind and the apiVersion fields in the hyperConverged CR. 

You can add them just above these lines; e.g.
```suggestion
	hc.APIVersion = "hco.kubevirt.io/v1beta1"
	hc.Kind = "HyperConverged"

	res := cli.RestClient().Put().
		Resource("hyperconvergeds").
		Name(hcov1beta1.HyperConvergedName).
		Namespace(flags.KubeVirtInstallNamespace).
		AbsPath("/apis", hcov1beta1.SchemeGroupVersion.Group, hcov1beta1.SchemeGroupVersion.Version).
		Timeout(10 * time.Second).
		Body(hc).Do(context.TODO())
```

File: tests/func-tests/virt-launcher-mutating-webhook.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
+package tests
```
Comment (nunnatsa) [inline tests/func-tests/virt-launcher-mutating-webhook.go:1]: Please rename the file to have the standard golang test file sufix `*_test.go`.

**Comment by iholder101** [comment]
This mechanism is now merged to Kubevirt: https://github.com/kubevirt/kubevirt/pull/9322.
We can close this PR.

/close

**Pull request closed without merging**